File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/screens/KeyfileExportScreen.kt
// "Back up your account key" — exports the UMK into a passphrase-
// encrypted `.flagshipkey` file. Reached from Settings → Recovery.
// Mirror of the iOS KeyfileExportScreen.
//
// Copy is approved + verbatim. The "Create backup file" CTA enables only
// when a strong passphrase is set + confirmed AND all three
// acknowledgments are checked. On success the file is saved via the
// Storage Access Framework (CreateDocument) as `<username>.flagshipkey`;
// we never write it ourselves.

package com.flagshipserver.app.ui.screens

import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckBox
import androidx.compose.material.icons.filled.CheckBoxOutlineBlank
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.navigation.NavController
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.ui.components.FSCard
import com.flagshipserver.app.ui.components.FSField
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import com.flagshipserver.app.viewmodels.KeyfileExportPhase
import com.flagshipserver.app.viewmodels.KeyfileExportViewModel

@Composable
fun KeyfileExportScreen(
    nav: NavController,
    // Onboarding seam — when set, a "Continue into the app" CTA appears
    // after the file is saved so the create-flow can proceed. Default is
    // a no-op so the Settings call site is unchanged.
    onSaved: (() -> Unit)? = null,
) {
    val app = LocalAppState.current
    val username = app.currentUser.collectAsState().value ?: ""
    val vm = remember(username) { KeyfileExportViewModel(username = username) }

    val phase by vm.phase.collectAsState()
    val passphrase by vm.passphrase.collectAsState()
    val confirm by vm.confirmPassphrase.collectAsState()
    val ackControl by vm.ackControl.collectAsState()

    // SAF "create document" — hands back a content:// uri we write the
    // keyfile text into. We never persist the file ourselves.
    var pendingText by remember { mutableStateOf<String?>(null) }
    var saved by remember { mutableStateOf(false) }
    val ctx = LocalContext.current
    val createDoc = rememberLauncherForActivityResult(
        ActivityResultContracts.CreateDocument("application/octet-stream"),
    ) { uri ->
        val text = pendingText
        if (uri != null && text != null) {
            runCatching {
                ctx.contentResolver.openOutputStream(uri)?.use { it.write(text.toByteArray(Charsets.UTF_8)) }
            }
            saved = true
        }
        pendingText = null
        vm.reset()
    }

    // When the wrap finishes, kick off the SAF save dialog.
    LaunchedEffect(phase) {
        (phase as? KeyfileExportPhase.Ready)?.let { ready ->
            pendingText = ready.text
            saved = false
            createDoc.launch(vm.suggestedFilename)
        }
    }

    val working = phase is KeyfileExportPhase.Working

    Column(
        Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = FS.space.s6)
            .semantics { contentDescription = "keyfile-export" },
    ) {
        Spacer(Modifier.height(FS.space.s8))
        FSGhostButton(label = "← Back", onClick = { nav.popBackStack() })
        Spacer(Modifier.height(FS.space.s3))
        Text(
            "ACCOUNT BACKUP",
            color = FS.colors.textMuted,
            style = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.SemiBold),
        )
        Spacer(Modifier.height(FS.space.s4))

        // Danger callout.
        FSCard(padding = PaddingValues(FS.space.s4)) {
            Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(FS.space.s2)) {
                Icon(Icons.Filled.Warning, contentDescription = null, tint = FS.colors.danger, modifier = Modifier.size(20.dp))
                Text(
                    "Anyone with both this file and its passphrase can take over your account and lock you out.",
                    color = FS.colors.text,
                    style = TextStyle(fontSize = 13.sp, lineHeight = 19.sp),
                )
            }
        }
        Spacer(Modifier.height(FS.space.s3))

        // Passphrase.
        FSCard(padding = PaddingValues(FS.space.s4)) {
            Column(verticalArrangement = Arrangement.spacedBy(FS.space.s3)) {
                FSField(
                    value = passphrase,
                    onValueChange = vm::setPassphrase,
                    label = "Passphrase",
                    helper = if (passphrase.isEmpty() || vm.passphraseStrong) "12 characters minimum" else null,
                    error = if (passphrase.isEmpty() || vm.passphraseStrong) null
                    else "12 characters minimum",
                    visualTransformation = PasswordVisualTransformation(),
                    modifier = Modifier.semantics { contentDescription = "keyfile-export-passphrase" },
                )
                FSField(
                    value = confirm,
                    onValueChange = vm::setConfirmPassphrase,
                    label = "Confirm Passphrase",
                    error = if (confirm.isEmpty() || vm.passphrasesMatch) null else "Passphrases don't match.",
                    visualTransformation = PasswordVisualTransformation(),
                    modifier = Modifier.semantics { contentDescription = "keyfile-export-confirm" },
                )
            }
        }
        Spacer(Modifier.height(FS.space.s3))

        // Acknowledgment.
        FSCard(padding = PaddingValues(FS.space.s4)) {
            Column(verticalArrangement = Arrangement.spacedBy(FS.space.s3)) {
                AckCheckbox(
                    checked = ackControl,
                    onToggle = { vm.setAckControl(!ackControl) },
                    text = "I understand anyone with this file and passphrase controls my entire account.",
                    cd = "keyfile-export-ack-control",
                )
            }
        }
        Spacer(Modifier.height(FS.space.s4))

        (phase as? KeyfileExportPhase.Failed)?.let { f ->
            FSCard(padding = PaddingValues(FS.space.s3)) {
                Text(f.message, color = FS.colors.danger, style = TextStyle(fontSize = 13.sp))
            }
            Spacer(Modifier.height(FS.space.s2))
        }

        FSPrimaryButton(
            label = if (working) "Creating…" else "Create backup file",
            onClick = { vm.createBackup() },
            enabled = vm.canCreate && !working,
            block = true,
            large = true,
            modifier = Modifier.semantics { contentDescription = "keyfile-export-create" },
        )

        if (saved) {
            Spacer(Modifier.height(FS.space.s3))
            FSCard(padding = PaddingValues(FS.space.s4)) {
                Row(verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(FS.space.s2)) {
                    Icon(Icons.Filled.CheckCircle, contentDescription = null, tint = FS.colors.success, modifier = Modifier.size(20.dp))
                    Text(
                        "Backup saved. Keep it somewhere safe and offline.",
                        color = FS.colors.text,
                        style = TextStyle(fontSize = 13.sp),
                        modifier = Modifier.semantics { contentDescription = "keyfile-export-saved" },
                    )
                }
            }
            if (onSaved != null) {
                Spacer(Modifier.height(FS.space.s3))
                FSPrimaryButton(
                    label = "Continue into the app",
                    onClick = onSaved,
                    block = true,
                    large = true,
                    modifier = Modifier.semantics { contentDescription = "keyfile-export-continue" },
                )
            }
        }
        Spacer(Modifier.height(FS.space.s8))
    }
}

@Composable
private fun AckCheckbox(checked: Boolean, onToggle: () -> Unit, text: String, cd: String) {
    Row(
        modifier = Modifier
            .clickable { onToggle() }
            .semantics { contentDescription = cd },
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(FS.space.s2),
    ) {
        Icon(
            if (checked) Icons.Filled.CheckBox else Icons.Filled.CheckBoxOutlineBlank,
            contentDescription = null,
            tint = if (checked) FS.colors.primary else FS.colors.textMuted,
            modifier = Modifier.size(22.dp),
        )
        Text(
            text,
            color = FS.colors.text,
            style = TextStyle(fontSize = 13.sp, lineHeight = 19.sp),
        )
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/Screens/KeyfileExportScreen.swift
import SwiftUI
import UniformTypeIdentifiers
import FlagshipCore

/// "Back up your account key" — exports the UMK into a passphrase-
/// encrypted `.flagshipkey` file. Reached from Settings → Recovery.
///
/// The "Create backup file" CTA enables only when a 12+-character
/// passphrase is set + confirmed and the control acknowledgment is checked.
/// On success the file is offered via the
/// share sheet as `<username>.flagshipkey`; we never write it ourselves.
public struct KeyfileExportScreen: View {
    @Environment(\.colorScheme) private var scheme
    @Bindable var vm: KeyfileExportViewModel

    @State private var shareURL: URL?
    @State private var showSaved = false

    public init(vm: KeyfileExportViewModel) {
        self.vm = vm
    }

    public var body: some View {
        let c = FSColors.scheme(scheme)
        ScrollView {
            VStack(alignment: .leading, spacing: FS.space.s4) {
                dangerCard(c: c)
                passphraseCard(c: c)
                acknowledgmentCard(c: c)

                if case .failed(let msg) = vm.phase {
                    ErrorCard(message: msg)
                }

                FSPrimaryButton(
                    "Create backup file",
                    enabled: vm.canCreate && !isWorking,
                    block: true,
                    large: true
                ) {
                    Task { await vm.createBackup() }
                }
                .accessibilityIdentifier("keyfile-export-create")

                if showSaved {
                    savedCard(c: c)
                }
            }
            .padding(FS.space.s6)
        }
        .background(c.bg.ignoresSafeArea())
        .navigationTitle("Account backup")
        .navigationBarTitleDisplayMode(.inline)
        .onChange(of: readyText) { _, text in
            guard let text else { return }
            shareURL = Self.writeTempKeyfile(text: text, filename: vm.suggestedFilename)
        }
        .sheet(item: $shareURL) { url in
            ShareSheet(items: [url]) {
                // After the share sheet dismisses, surface the
                // "saved" confirmation + clean up the temp file.
                showSaved = true
                vm.reset()
                try? FileManager.default.removeItem(at: url)
            }
        }
    }

    private var isWorking: Bool {
        if case .working = vm.phase { return true }
        return false
    }

    private var readyText: String? {
        if case .ready(let text) = vm.phase { return text }
        return nil
    }

    private func dangerCard(c: FSColors) -> some View {
        FSCard {
            HStack(alignment: .top, spacing: FS.space.s2) {
                Image(systemName: "exclamationmark.triangle.fill")
                    .foregroundColor(c.danger)
                Text("Anyone with both this file and its passphrase can take over your account and lock you out.")
                    .font(FS.font.bodySm())
                    .foregroundColor(c.text)
            }
        }
        .accessibilityIdentifier("keyfile-export-danger")
    }

    private func passphraseCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                FSField(
                    value: $vm.passphrase,
                    label: "Passphrase",
                    helper: "12 characters minimum",
                    error: vm.passphrase.isEmpty || vm.passphraseStrong
                        ? nil
                        : "12 characters minimum",
                    secure: true
                )
                .accessibilityIdentifier("keyfile-export-passphrase")
                FSField(
                    value: $vm.confirmPassphrase,
                    label: "Confirm Passphrase",
                    error: vm.confirmPassphrase.isEmpty || vm.passphrasesMatch
                        ? nil
                        : "Passphrases don't match.",
                    secure: true
                )
                .accessibilityIdentifier("keyfile-export-confirm")
            }
        }
    }

    private func acknowledgmentCard(c: FSColors) -> some View {
        FSCard {
            VStack(alignment: .leading, spacing: FS.space.s3) {
                checkbox(
                    isOn: $vm.ackControl,
                    text: "I understand anyone with this file and passphrase controls my entire account.",
                    id: "keyfile-export-ack-control",
                    c: c
                )
            }
        }
    }

    private func savedCard(c: FSColors) -> some View {
        FSCard {
            HStack(alignment: .top, spacing: FS.space.s2) {
                Image(systemName: "checkmark.seal.fill").foregroundColor(c.success)
                Text("Backup saved. Keep it somewhere safe and offline.")
                    .font(FS.font.bodySm())
                    .foregroundColor(c.text)
            }
        }
        .accessibilityIdentifier("keyfile-export-saved")
    }

    private func checkbox(isOn: Binding<Bool>, text: String, id: String, c: FSColors) -> some View {
        Button(action: { isOn.wrappedValue.toggle() }) {
            HStack(alignment: .top, spacing: FS.space.s2) {
                Image(systemName: isOn.wrappedValue ? "checkmark.square.fill" : "square")
                    .foregroundColor(isOn.wrappedValue ? c.primary : c.textMuted)
                    .imageScale(.large)
                Text(text)
                    .font(FS.font.bodySm())
                    .foregroundColor(c.text)
                    .multilineTextAlignment(.leading)
                Spacer(minLength: 0)
            }
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(id)
        .accessibilityAddTraits(isOn.wrappedValue ? [.isSelected] : [])
    }

    /// Write the keyfile text to a temp file named
    /// `<username>.flagshipkey` so the share sheet hands the user a real
    /// file with the right name + extension (rather than raw text).
    static func writeTempKeyfile(text: String, filename: String) -> URL? {
        let dir = FileManager.default.temporaryDirectory
        let url = dir.appendingPathComponent(filename)
        try? FileManager.default.removeItem(at: url)
        do {
            try text.data(using: .utf8)?.write(to: url, options: [.atomic, .completeFileProtection])
            return url
        } catch {
            return nil
        }
    }
}

extension URL: @retroactive Identifiable {
    public var id: String { absoluteString }
}

/// UIActivityViewController bridge for SwiftUI — the share sheet used to
/// hand the user the `.flagshipkey` file (AirDrop, Files, etc.).
struct ShareSheet: UIViewControllerRepresentable {
    let items: [Any]
    var onDismiss: () -> Void = {}

    func makeUIViewController(context: Context) -> UIActivityViewController {
        let vc = UIActivityViewController(activityItems: items, applicationActivities: nil)
        vc.completionWithItemsHandler = { _, _, _, _ in onDismiss() }
        return vc
    }

    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}
